R/qCount.R

```diff
@@ -3,516 +3,895 @@
-                               includeSecondary=includeSecondary,
-                               mapqmin=as.integer(mapqMin)[1],
-                               mapqmax=as.integer(mapqMax)[1],
-                               absisizemin=as.integer(absIsizeMin)[1],
-                               absisizemax=as.integer(absIsizeMax)[1]))
+        ## apply 'mask' to flatquery -------------------------------------------
+        if (!is.null(mask)) {
+            if (!inherits(mask, "GRanges"))
+                stop("'mask' must be an object of type 'GRanges'")
+            message("removing 'mask' ranges from 'query'...", appendLF = FALSE)
+            BiocGenerics::strand(mask) <- "*"
+            mask <- GenomicRanges::reduce(mask)
+            ov <- GenomicRanges::findOverlaps(flatquery, mask)
+            qOM <- unique(S4Vectors::queryHits(ov))
+            gr1 <- GenomicRanges::GRanges(
+                seqnames = S4Vectors::Rle(qOM), 
+                ranges = IRanges::IRanges(start = BiocGenerics::start(flatquery)[qOM], 
+                                          end = BiocGenerics::end(flatquery)[qOM])
+            )
+            gr2 <- GenomicRanges::GRanges(
+                seqnames = S4Vectors::Rle(S4Vectors::queryHits(ov)), 
+                ranges = IRanges::IRanges(
+                    start = BiocGenerics::start(mask)[S4Vectors::subjectHits(ov)], 
+                    end = BiocGenerics::end(mask)[S4Vectors::subjectHits(ov)]
+                )
+            )
+            SD <- BiocGenerics::setdiff(gr1, gr2)
+            notOverlappingMaskInd <- which(!((seq_len(length(flatquery))) %in% qOM))
+            completelyMaskedInd <- 
+                qOM[!(qOM %in% as.numeric(as.character(unique(GenomicRanges::seqnames(SD)))))]
+            
+            # 'zeroquery' contains regions that are completely masked (will get zero count and length)
+            #zeroquery <- flatquery[completelyMaskedInd]
+            tmpnames <- querynames[completelyMaskedInd]
+            zeroquerynames <- c(zeroquerynames, tmpnames[!(tmpnames %in% querynames[-completelyMaskedInd])])
+            #zeroquerylengths <- rep(0L, length(completelyMaskedInd))
+            
+            # masked 'flatquery' maybe split into several non-masked pieces
+            flatquery <- c(flatquery[notOverlappingMaskInd],
+                           GenomicRanges::GRanges(
+                               seqnames = GenomicRanges::seqnames(flatquery)[as.numeric(as.character(GenomicRanges::seqnames(SD)))],
+                               ranges = IRanges::ranges(SD), 
+                               strand = GenomicRanges::strand(flatquery)[as.numeric(as.character(GenomicRanges::seqnames(SD)))],
+                               seqlengths = GenomeInfoDb::seqlengths(flatquery)),
+                           ignore.mcols = TRUE)
+            querynames <- querynames[c(notOverlappingMaskInd, as.numeric(as.character(GenomicRanges::seqnames(SD))))]
```

Comment:
`SD` is a `GRanges` and we imported the method `GenomicRanegs::seqnames` at the top of this function, but as I commented above, I am not sure if we should import the generic `GenomeInfoDB::seqnames` and use that here in this call. They both seem to work and give identical return values when called on a toy GRanges object.